sdk-base/src/main/java/com/mapbox/maps/plugin/viewport/experimental/ViewportPlugin.kt

```diff
@@ -0,0 +1,154 @@
+   *
+   * When started, goes to [ViewportTransition]
+   * and to the final [ViewportState] when ended.
+   *
+   * If transition is canceled, state goes to IDLE.
+   *
+   * @param targetState The target [ViewportState] to transition to.
+   * @param completionListener The listener to observe the completion state.
+   */
+  fun transitionTo(targetState: ViewportState, completionListener: CompletionListener? = null)
+
+  /**
+   * Immediately goes to IDLE state canceling all ongoing transitions.
+   */
+  fun transitionToIdle()
+
+  /**
+   * Indicates whether to transition [ViewportState] to IDLE (null) when user interact with the map
+   * using gestures.
+   *
+   * Defaults to true.
+   */
+  var transitionsToIdleUponUserInteraction: Boolean
+
+  // States
+
+  /**
+   * Returns list of registered states.
+   */
+  val states: Set<ViewportState>
+
+  /**
+   * Add a viewport state to the viewport plugin, which could be reused later with the state id.
+   *
+   * @param state the view port state to be added.
+   */
+  fun addState(state: ViewportState)
+
+  /**
+   * Remove a viewport state from the viewport plugin, and it could not be reused later with the state id.
+
+   * @param state the view port state to be removed.
+   */
+  fun removeState(state: ViewportState)
+
+  // Transitions
+
+  /**
+   * DefaultViewportTransition with default options
+   */
+  var defaultTransition: ViewportTransition
+
+  /**
+   * Set the [ViewportTransition] for the transition from given [ViewportState] to target [ViewportState]
+   *
+   * @param transition The transition to be set.
+   * @param from The state before the transition.
+   * @param to The state after the transition.
+   */
+  fun setTransition(transition: ViewportTransition, from: ViewportState?, to: ViewportState)
+
+  /**
+   * Get the transition from the previous state and target state.
+   *
+   * @param from The state before the transition.
+   * @param to The state after the transition.
+   */
+  fun getTransition(from: ViewportState?, to: ViewportState): ViewportTransition?
+
+  /**
+   * Remove the transition between the previous state and target state.
+   *
+   * @param from The state before the transition.
+   * @param to The state after the transition.
+   */
+  fun removeTransition(from: ViewportState?, to: ViewportState)
+
+  // Observers
+
+  /**
+   * Adds [ViewportStatusObserver].
+   */
+  fun addStatusObserver(
+    viewportStatusObserver: ViewportStatusObserver
+  )
+
+  /**
+   * Removes [ViewportStatusObserver].
+   */
+  fun removeStatusObserver(
+    viewportStatusObserver: ViewportStatusObserver
+  )
+
+  // Convenient methods to create the in-stock [ViewportState] and [ViewportTransition].
+
+  /**
+   * Create a [FollowingViewportState] instance with provided [FollowingViewportStateOptions].
+   *
+   * @param options The desired [FollowingViewportStateOptions]
+   */
+  fun makeFollowingViewportState(options: FollowingViewportStateOptions): FollowingViewportState
```

Comment:
the `make…` naming is a Swift convention for factory methods. If there's a Kotlin convention, we should probably use that instead on Android.

Reply:
I don't think there's a Kotlin convention, we could probably use `build` or `create`.